restorer_test.go

```diff
@@ -189,89 +212,208 @@ func testRestorer(t *testing.T, when spec.G, it spec.S) {
+				})
 			})
 
-			it("escapes buildpack IDs when restoring buildpack layers", func() {
-				h.AssertNil(t, restorer.Restore(testCache))
-				expectedMetadata := `[metadata]
+			when("there is a cache=true escaped layer", func() {
+				var meta string
+				it.Before(func() {
+					meta = `cache=true
+[metadata]
   escaped-bp-key = "escaped-bp-val"`
-				if txt, err := ioutil.ReadFile(filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer.toml")); err != nil {
-					t.Fatalf("failed to read escaped-bp-layer.toml: %s", err)
-				} else if !strings.Contains(string(txt), expectedMetadata) {
-					t.Fatalf(`Error: expected '%s' to contain '%s'`, txt, expectedMetadata)
-				}
-
-				expectedText := "echo text from escaped bp layer"
-				if txt, err := ioutil.ReadFile(filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer", "file-from-escaped-bp")); err != nil {
-					t.Fatalf("failed to read file-from-escaped-bp: %s", err)
-				} else if !strings.Contains(string(txt), expectedText) {
-					t.Fatalf(`Error: expected '%s' to contain '%s'`, txt, expectedText)
-				}
+					h.AssertNil(t, writeLayer(layersDir, "escaped_buildpack_id", "escaped-bp-layer", meta, escapedLayerSHA))
+					h.AssertNil(t, restorer.Restore(testCache))
+				})
+
+				it("keeps layer metadatata", func() {
+					got := h.MustReadFile(t, filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer.toml"))
+					h.AssertEq(t, string(got), meta)
+				})
+				it("keeps layer sha", func() {
+					got := h.MustReadFile(t, filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer.sha"))
+					h.AssertEq(t, string(got), escapedLayerSHA)
+				})
+				it("restores data", func() {
+					got := h.MustReadFile(t, filepath.Join(layersDir, "escaped_buildpack_id", "escaped-bp-layer", "file-from-escaped-bp"))
+					want := "echo text from escaped bp layer\n"
+					h.AssertEq(t, string(got), want)
+				})
+			})
+
+			when("there is a cache=true an unknown buildpack", func() {
+				it.Before(func() {
+					meta := "cache=true"
+					h.AssertNil(t, writeLayer(layersDir, "some.unknown.buildpack.id", "cache-only", meta, cacheOnlyLayerSHA))
+					h.AssertNil(t, restorer.Restore(testCache))
+				})
+
+				it("keeps metadata and sha file", func() {
+					h.AssertPathExists(t, filepath.Join(layersDir, "some.unknown.buildpack.id", "cache-only.toml"))
+					h.AssertPathExists(t, filepath.Join(layersDir, "some.unknown.buildpack.id", "cache-only.sha"))
+				})
+				it("does not restore layer data", func() {
+					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "some.unknown.buildpack.id", "cache-only"))
+				})
+			})
+
+			when("there is a cache=true layer in cache but not in group", func() {
+				it.Before(func() {
+					meta := "cache=true"
+					h.AssertNil(t, writeLayer(layersDir, "nogroup.buildpack.id", "some-layer", meta, noGroupLayerSHA))
+					h.AssertNil(t, restorer.Restore(testCache))
+				})
+
+				it("keeps metadata and sha file", func() {
```

Comment:
Same as above

Reply:
Removed the 'keeps metadata' part but kept the 'does not restore data', because I think that's still relevant.